File: 700/763 - Partition Labels.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> partitionLabels(String S) {
        List<Integer> ret = new ArrayList<>();
        int len = S.length();
        int[] last = new int[26];
        int p = 0, l = 0;
        
        for (int i = 0; i < len; ++i) {
            last[S.charAt(i) - 'a'] = i;
        }
        
        for (int i = 0; i < len; ++i) {
            l = Math.max(l, last[S.charAt(i) - 'a']);
            
            if (i == l) {
                ret.add(i - p + 1);
                p = i + 1;
            }
        }
        
        return ret;
    }
}

File: csharp/763.partition-labels.cs
/*
 * @lc app=leetcode id=763 lang=csharp
 *
 * [763] Partition Labels
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public IList<int> PartitionLabels(string S) {
        int n = S.Length, start = 0, end = 0;
        var map = new Dictionary<char, int>();
        var ret = new List<int>();
        for (int i = 0; i < n; ++i) {
            map[S[i]] = i;
        }
        for (int i = 0; i < n; ++i) {
            end = Math.Max(map[S[i]], end);
            if (i == end) {
                ret.Add(end - start + 1);
                start = end + 1;
                end = start;
            }
        }
        return ret;
    }
}
// @lc code=end

